Add Unordered Sets + fix maps iteration section (Java & C++) (#2390)
cpinitiative/usaco-guide: content/2_Bronze/Intro_Sets.mdx

</Resources>

Both Java and C++ contain two versions of sets and maps; one using **sorting**

> and the other using **hashing**. For Python, we'll introduce the built-in sets

Reviewer: maybe put this stuff in language sections

<CPPSection>

### Unordered Sets

In _unordered_ sets, elements are stored in a random order. Insertions,
deletions, and searches are all $\mathcal{O}(1)$. Unordered sets are implemented
by
[`std:unordered_set`](https://www.cplusplus.com/reference/unordered_set/unordered_set/)
in the `<unordered_set>` header.

Some operations on an `std::unordered_set` named `s` include:

- `s.insert(x)`, which adds the element `x` to `s` if not already present.
- `s.erase(x)`, which removes the element `x` from `s` if present.
- `s.count(x)`, which returns `1` if `s` contains `x` and `0` if it doesn't.

```cpp
unordered_set<int> s;
s.insert(1);  // {1}
s.insert(4);  // {1, 4}
s.insert(2);  // {1, 4, 2}

> s.insert(1);  // {1, 4, 2}
> // the insert method did nothing because 1 was already in the set

Reviewer: ```suggestion
s.insert(1);  // does nothing because 1's already in the set
```

cout << s.count(1) << endl;  // 1

> s.erase(1);  // [2, 4]
> cout << s.count(5) << endl;  // 0
> s.erase(0);  // [2, 4]
> // if the element to be erased was not present, nothing happens

Reviewer: ```suggestion
s.erase(1);  // {2, 4}
cout << s.count(5) << endl;  // 0
s.erase(0);  // does nothing because 0 wasn't in the set
```

```

### Sorted Sets

In _sorted_ sets, the elements are sorted in order of element. Insertions,
deletions, and searches are all $\mathcal{O}(\log N)$, where $N$ is the number
of elements in the set. Sorted sets are implemented by
[`std:set`](https://www.cplusplus.com/reference/set/set/) in the `<set>`
header.

`std::set` includes all the essential operations that `std::unordered_set` has
(including insertion, deletion, searches, etc.), but also some additional ones.
Refer to the
[More Operations on Sorted Sets](https://usaco.guide/silver/intro-sorted-sets)
module for more detail.

You can iterate through a set in sorted order using a for-each loop.

```cpp
set<int> s;
s.insert(1);  // [1]
s.insert(4);  // [1, 4]

> s.insert(2);  // [1, 2, 4]

Reviewer: ```suggestion
s.insert(1);  // {1}
s.insert(4);  // {1, 4}
s.insert(2);  // {1, 2, 4}

Author: actually imma revert this since i think the point of square brackets is to emphasize that the set is ordered (this is the same everywhere else)

// Outputs 1, 2, and 4 on separate lines
for (int element : s) {
	cout << element << endl;
}
```

</CPPSection>

<JavaSection>

### Unordered Sets

In _unordered_ sets, elements are stored in a random order. Insertions,
deletions, and searches are all $\mathcal{O}(1)$. Unordered sets are implemented
in Java by the
[`HashSet`](https://docs.oracle.com/javase/7/docs/api/java/util/HashSet.html)
class (which comes from the `java.util` library).

Some operations on a `HashSet` named `set` include:

- `set.add(x)`, which adds the element `x` to `set` if not already present.
- `set.remove(x)`, which removes the element `x` from `set` if present.
- `set.contains(x)`, which checks whether `set` contains the element `x`.

```java
Set<Integer> set = new HashSet<>();
set.add(1);  // {1}
set.add(4);  // {1, 4}
set.add(2);  // {1, 4, 2}
set.add(1);  // {1, 4, 2}
// the add method did nothing because 1 was already in the set
System.out.println(set.contains(1));  // true
set.remove(1);  // {4, 2}
System.out.println(set.contains(5));  // false

> set.remove(0);  // {4, 2}
> // if the element to be removed does not exist, nothing happens

Reviewer: ```suggestion
set.remove(0);  // does nothing because 0 wasn't in the set
```

<JavaSection>

### Unordered Sets

In _unordered_ sets, elements are stored in a random order. Insertions,
deletions, and searches are all $\mathcal{O}(1)$, where $N$ is the number of
elements in the set. Unordered sets are implemented in Java by the `HashSet`
class.

Some operations on a `HashSet` named `set` include:

- `set.add(x)`, which adds the element `x` to `set` if not already present.
- `set.remove(x)`, which removes the element `x` from `set` if present.
- `set.contains(x)`, which checks whether `set` contains the element `x`.

```java
Set<Integer> set = new HashSet<>();

> set.add(1); // {1}

Reviewer: two spaces after code before starting `//`

Reviewer: not 100% addressed, some other places too

Reviewer: not 100% addressed, some other places too

the set. Sorted sets are implemented in Java by the `TreeSet` class.

`TreeSet` includes all the operations that `HashSet` has, but also includes some
additional ones. Some useful ones are listed below (all $\mathcal{O}(\log N)$):

- `first()`, which returns the smallest element in the set and `last()`, which
  returns the largest element.
- `lower(x)`, which returns the largest element strictly smaller than `x` and
  `higher(x)`, which returns the smallest element strictly larger than `x`.
- `floor(x)`, which returns the largest element that's less than or equal to `x`
  and `ceiling(x)`, which returns the smallest element that's greater than or
  equal to `x`.

You can also iterate through a `TreeSet` in sorted order using a for-each loop.

```java
Set<Integer> set = new TreeSet<>();
set.add(1); // {1}
set.add(4); // {1, 4}
set.add(2); // {1, 2, 4}

System.out.println(set.lower(3)); // 2
System.out.println(set.higher(4)); // null
System.out.println(set.ceiling(4)); // 4
System.out.println(set.floor(3)); // 2

// You can iterate through a TreeSet set in sorted order
for (int element : set) {
	System.out.println(element);
}

> // Outputs 1, 2, and 4 on separate lines

Reviewer: put above the for loop

In _sorted_ sets, the elements are sorted in order of element. Insertions,
deletions, and searches are all $\mathcal{O}(\log N)$, where $N$ is the number

> of elements in the set. Unordered sets are implemented by

Reviewer: unordered sets?

Author: good catch whoops

System.out.println(set.contains(5));  // false
set.remove(0);  // {4, 2}
// if the element to be removed does not exist, nothing happens
```

### Sorted Sets

In _sorted_ sets, the elements are stored in order. Insertions, deletions, and
searches are all $\mathcal{O}(\log N)$, where $N$ is the number of elements in
the set. Sorted sets are implemented in Java by the `TreeSet` class.

`TreeSet` includes all the operations that `HashSet` has, but also includes some
additional ones. Some useful ones are listed below (all $\mathcal{O}(\log N)$):

- `first()`, which returns the smallest element in the set and `last()`, which
  returns the largest element.
- `lower(x)`, which returns the largest element strictly smaller than `x` and
  `higher(x)`, which returns the smallest element strictly larger than `x`.
- `floor(x)`, which returns the largest element that's less than or equal to `x`
  and `ceiling(x)`, which returns the smallest element that's greater than or
  equal to `x`.

You can also iterate through a `TreeSet` in sorted order using a for-each loop.

```java
Set<Integer> set = new TreeSet<>();
set.add(1);  // {1}
set.add(4);  // {1, 4}
set.add(2);  // {1, 2, 4}

> System.out.println(set.lower(3)); // 2
> System.out.println(set.higher(4)); // null
> System.out.println(set.ceiling(4)); // 4
> System.out.println(set.floor(3)); // 2

Reviewer: ```suggestion
System.out.println(set.lower(3));  // 2
System.out.println(set.higher(4));  // null
System.out.println(set.ceiling(4));  // 4
System.out.println(set.floor(3));  // 2
```

<JavaSection>

To iterate over maps, you can use a for-each loop over the keys:

```java
for (int k : m.keySet()) {
	System.out.println(k + " " + m.get(k));
}
```

You can also use a for-each loop over the entries:

```java

> for (Map.Entry entry : m.entrySet()) {

Reviewer: oh uh
(optional) mention somewhere in java & c++ that like all these set/dict things are in `java.util` & `<set>` + `<map>` respectively

Author: I'm not sure what to say in c++ since I'm not too familiar with it, could you write out what that would look like?

Reviewer: uh
in c++ `std::set` is in the header `<set>` and `std::map` is in the header `<map>`

Author: So like would I say unordered sets are implemented by std::unordered_set in the <set> header?

Reviewer: uh unordered sets are actually in `<unordered_set>` lol

Reviewer: don't see you mentioning `java.util` anywhere?

Author: mentioned on line 121

Reviewer: i'll bellieve you

```java
Map<Integer, Integer> m = new TreeMap<>();
// m starts as {0: 0, 1: 1, 2: 2}

> m.put(0, 0); m.put(1, 1); m.put(2, 2);

Reviewer: statements each on own line

Author: is there a need? i feel like this saves space and leaves room for the more important stuff. it also makes sense to group the operations on one line

Reviewer: hm
good point, actually

elements while looping over them, like in the next code snippet:

```java
Map<Integer, Integer> m = new TreeMap<>();
// m starts as {0: 0, 1: 1, 2: 2}
m.put(0, 0); m.put(1, 1); m.put(2, 2);

Iterator<Map.Entry<Integer, Integer>> iter = m.entrySet().iterator();
while (iter.hasNext()) {
	int key = iter.next();
	if (key == 0 || key == 2) {
		iter.remove();
	}
}

// m ends as {1: 1}
System.out.println("Remaining Entries:");
for (Map.Entry<Integer, Integer> entry : m.entrySet()) {
	System.out.println(entry.getKey() + " " + entry.getValue());
}

/*
Remaining Entries:
1 1

> */

Reviewer: ok idk how a suggestion would work but how about just
```java
// The only remaining entry will be {1, 1}
System.out.println("Remaining Entries:");
for (Map.Entry<Integer, Integer> entry : m.entrySet()) {
	System.out.println(entry.getKey() + " " + entry.getValue());
}
```

to the code above):

```java
Map<Integer, Integer> s = new TreeMap<>();
// m starts as {0: 0, 1: 1, 2: 2}

> m.put(0, 0); m.put(1, 1); m.put(2, 2);

Reviewer: same thing here

Set<Integer> keysToRemove = new TreeSet<>();
for (Map.Entry<Integer, Integer> entry : m.entrySet()) {
	int key = entry.getKey();
	if (key == 0 || key == 2) {
		keysToRemove.add(key);
	}
}

// removes all entries with key in keysToRemove
m.keySet().removeAll(keysToRemove);

// s ends as {1}
System.out.println("Remaining Entries:");
for (Map.Entry<Integer, Integer> entry : m.entrySet()) {
	System.out.println(entry.getKey() + " " + entry.getValue());
}

/*
Entries:

> 1 1

Reviewer: put as single line comment maybe

Author: i think the point is to show what the code will output

Reviewer: if so then at least format the multiline acceptably
```cpp
/*
 * for reference
 * something like this
 */
```

Reviewer: wait you alr did it, sorry

to the code above):

```java
Map<Integer, Integer> s = new TreeMap<>();
// m starts as {0: 0, 1: 1, 2: 2}
m.put(0, 0); m.put(1, 1); m.put(2, 2);

Set<Integer> keysToRemove = new TreeSet<>();
for (Map.Entry<Integer, Integer> entry : m.entrySet()) {
	int key = entry.getKey();
	if (key == 0 || key == 2) {
		keysToRemove.add(key);
	}
}

// removes all entries with key in keysToRemove
m.keySet().removeAll(keysToRemove);

// s ends as {1}
System.out.println("Remaining Entries:");
for (Map.Entry<Integer, Integer> entry : m.entrySet()) {
	System.out.println(entry.getKey() + " " + entry.getValue());
}

/*
 * Remaining Entries:

>  * 1 1

Reviewer: same thing here

      </Quiz.Explanation>
    </Quiz.Answer>
    <Quiz.Answer correct>
      `false`
      <Quiz.Explanation>
        Correct. $3$ isn't present in `s`.
      </Quiz.Explanation>
    </Quiz.Answer>
    <Quiz.Answer>
      `null`
      <Quiz.Explanation>

      </Quiz.Explanation>
    </Quiz.Answer>
    <Quiz.Answer>
      Error: `NullPointerException`
      <Quiz.Explanation>

      </Quiz.Explanation>
    </Quiz.Answer>

  </Quiz.Question>
  <Quiz.Question>
    Which of the following is the correct output of this code?
  ```java
  Set<Integer> s = new TreeSet<>(Arrays.asList(4, 1, 2, 1, 3, 2, 3, 1, 3, 1));
  System.out.println(s.size());
  ```
    <Quiz.Answer>
      $1$
      <Quiz.Explanation>

      </Quiz.Explanation>
    </Quiz.Answer>
    <Quiz.Answer correct>
      $4$
      <Quiz.Explanation>
        Correct. There are $4$ unique elements in `s`.
      </Quiz.Explanation>
    </Quiz.Answer>
    <Quiz.Answer>
      $7$
      <Quiz.Explanation>

      </Quiz.Explanation>
    </Quiz.Answer>
    <Quiz.Answer>
      $10$
      <Quiz.Explanation>
    Sets only store unique elements.
      </Quiz.Explanation>
    </Quiz.Answer>

  </Quiz.Question>
  <Quiz.Question>

>     What is the time complexity of insertions, deletions, and removals in an unordered map of size $N$?

Reviewer 2: aren't removals and deletions the same?? maybe `insertions, deletions, and searches`

    <Quiz.Answer correct>
      $\mathcal{O}(1)$
      <Quiz.Explanation>
        Correct.
      </Quiz.Explanation>
    </Quiz.Answer>
    <Quiz.Answer>
      $\mathcal{O}(N)$
    </Quiz.Answer>
    <Quiz.Answer>
      $\mathcal{O}(N \log N)$
    </Quiz.Answer>

>     <Quiz.Answer correct>

Reviewer 2: ```suggestion
    <Quiz.Answer>
```

  </Quiz.Question>
</Quiz>
</CPPSection>

<JavaSection>
<Quiz>
  <Quiz.Question>
    What is the time complexity of insertions, deletions, and searches in a sorted set of size $N$?
    <Quiz.Answer>
      $\mathcal{O}(1)$
    </Quiz.Answer>
    <Quiz.Answer>
      $\mathcal{O}(N \log N)$
    </Quiz.Answer>
    <Quiz.Answer>
      $\mathcal{O}(N)$
    </Quiz.Answer>
    <Quiz.Answer correct>
      $\mathcal{O}(\log N)$
      <Quiz.Explanation>
		Correct.
      </Quiz.Explanation>
    </Quiz.Answer>
  </Quiz.Question>
  <Quiz.Question>
    Which of the following is the correct output of this code?

> 		```java

Author: 👀

		Set<Integer> s = new TreeSet<>();
		s.insert(1);
		s.insert(4);
		s.insert(5);
		s.insert(2);
		System.out.println(s.contains(3));
		```
    <Quiz.Answer>
      `true`
      <Quiz.Explanation>

      </Quiz.Explanation>
    </Quiz.Answer>
    <Quiz.Answer correct>
      `false`
      <Quiz.Explanation>
        Correct. $3$ isn't present in `s`.
      </Quiz.Explanation>
    </Quiz.Answer>
    <Quiz.Answer>
      `null`
      <Quiz.Explanation>

      </Quiz.Explanation>
    </Quiz.Answer>
    <Quiz.Answer>
      Error: `NullPointerException`
      <Quiz.Explanation>

      </Quiz.Explanation>
    </Quiz.Answer>

  </Quiz.Question>
  <Quiz.Question>
    Which of the following is the correct output of this code?
		```java

> 		Integer[] arr = {4, 1, 2, 1, 3, 2, 3, 1, 3, 1};

Reviewer: why `Integer[]`
that might confuse some people

Author: it doesn't work if it's int because java expects an object array

Reviewer: why not
```java
Set<Integer> set = new TreeSet<>(Arrays.asList(4, 1, 2, 1, 3, 2, 3, 1, 3, 1));
```

Author: oh wow didn't know you could do that
